plugins/microservices/src/json_compare.cpp

```diff
@@ -0,0 +1,189 @@
+#include "irods/irods_ms_plugin.hpp"
+#include "irods/irods_re_structs.hpp"
+#include "irods/msParam.h"
+#include "irods/objInfo.h" // For StrArray
+#include "irods/rodsErrorTable.h"
+#include "irods/storage.hpp"
+
+#include <boost/any.hpp>
+#include <nlohmann/json.hpp>
+
+#include <cstdint>
+#include <cstring>
+#include <functional>
+#include <string>
+#include <vector>
+
+namespace
+{
+    // clang-format off
+    namespace is  = irods::experimental::storage;
+
+    using log_msi = irods::experimental::log::microservice;
+    using json    = nlohmann::json;
+    // clang-format on
+
+    auto msi_impl(MsParam* _json_handle_1,
+                  MsParam* _json_pointer_1, // NOLINT(bugprone-easily-swappable-parameters)
+                  MsParam* _json_handle_2,
+                  MsParam* _json_pointer_2, // NOLINT(bugprone-easily-swappable-parameters)
+                  MsParam* _cmp_result,
+                  ruleExecInfo_t* _rei) -> int
+    {
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_handle_1);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_handle_2);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_pointer_1);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_pointer_2);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_cmp_result);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_rei);
+
+        IRODS_MSI_REQUIRE_TYPE(_json_handle_1->type, STR_MS_T);
+        IRODS_MSI_REQUIRE_TYPE(_json_handle_2->type, STR_MS_T);
+        IRODS_MSI_REQUIRE_TYPE(_json_pointer_1->type, STR_MS_T);
+        IRODS_MSI_REQUIRE_TYPE(_json_pointer_2->type, STR_MS_T);
+
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_handle_1->inOutStruct);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_handle_2->inOutStruct);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_pointer_1->inOutStruct);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_pointer_2->inOutStruct);
+
+        try {
+            auto* handle_1 = static_cast<char*>(_json_handle_1->inOutStruct);
+            log_msi::debug("First JSON handle (address=[{}], value=[{}])", static_cast<void*>(handle_1), handle_1);
+
+            auto* object = is::find(handle_1);
+            if (!object) { // NOLINT(readability-implicit-bool-conversion)
+                log_msi::error("Invalid argument: JSON handle [{}] is not associated with any object.", handle_1);
+                return SYS_INVALID_INPUT_PARAM;
+            }
+
+            const auto& json_object_1 = boost::any_cast<json&>(*object);
+            const json* json_value_1{};
+
+            try {
+                const auto* p = static_cast<const char*>(_json_pointer_1->inOutStruct);
+                // NOLINTNEXTLINE(readability-implicit-bool-conversion)
+                log_msi::debug("First JSON pointer (address=[{}], value=[{}])", static_cast<const void*>(p), (p ? p : "<nullptr>"));
+
+                // TODO Should this result in an error if the the JSON pointer is invalid?
+                json_value_1 = &json_object_1.at(json::json_pointer{p});
```

Comment:
See TODO comment.

Reply:
I'm going to make it so that the msi returns an error if any of the inputs are invalid.
- `INVALID_HANDLE` (a new generic error code) will be returned if either handle is invalid
- `DOES_NOT_EXIST` (a new generic error code) will be returned if either json pointer is invalid

This change will apply to all JSON microservices.